python/ray/experimental/data/tests/test_dataset.py

```diff
@@ -274,6 +276,90 @@ def test_read_binary_files_with_fs(ray_start_regular_shared):
             assert expected == item
 
 
+def test_iter_batches(ray_start_regular_shared):
```

Comment:
Can you add a brute force grid search test that iterating across all rows returns correct answers for range(20) and batch size from 0..20., prefetch from 0..20?

This will help catch correctness bugs in the slicing, which seems pretty complex.

Reply:
I can look at doing that, here's why I initially opted to not do that:
1. The block prefetching logic and batch size logic are independent, since we process (slice) one block at a time (not all prefetched blocks) along with a possibly empty remainder from the previous block. I.e., the block prefetching logic doesn't depend on the batch size, and the batch size logic doesn't depend on the number of prefetched blocks.
2. The core batch size cases are already covered: (a) batch size equal to block size, (b) batch size smaller than block size, (c) batch size greater than block size, (d) batch size not a divisor of block size and drop partial, (e) batch size not a divisor and not drop partial. Other cases that I thought about covering and could still add: (f) batch size an unequal divisor of block size (e.g. `batch_size=2`, `block_size=4`), (g) batch size 2x block size.

If we went the grid route, here's how I think we could generically validate the slicing:
```python
# Grid: num_blocks x num_rows_per_block x batch_size x drop_last
for num_blocks in range(1, 10):
    for num_rows_per_block in range(1, 10):
        for batch_size in range(1, num_blocks * num_rows_per_block):
            for drop_last in [False, True]:
                ds = ray.experimental.data.range_arrow(
                    num_blocks * num_rows_per_block, parallelism=num_blocks)
                batches = list(ds.iter_batches(batch_size=batch_size, drop_last=drop_last))
                if num_blocks * num_rows_per_block % batch_size == 0 or not drop_last:
                    assert len(batches) == math.ceil(num_blocks * num_rows_per_block / batch_size)
                    assert pd.concat(batches, ignore_index=True).equals(pd.concat(ray.get(ds.to_pandas())))
                else:
                    assert len(batches) == num_blocks * num_rows_per_block // batch_size
                    assert pd.concat(batches, ignore_index=True).equals(pd.concat(ray.get(ds.to_pandas()))[:batch_size * (num_blocks * num_rows_per_block // batch_size)])
                if num_blocks * num_rows_per_block % batch_size == 0 or drop_last:
                    assert all(len(batch) == batch_size for batch in batches)
                else:
                    assert all(len(batch) == batch_size for batch in batches[:-1])
                    assert len(batches[-1]) == num_blocks * num_rows_per_block % batch_size
```

Reply:
I'm taking a stab at this right now.